base/compiler/ssair/passes.jl

```diff
@@ -2355,6 +2351,20 @@ function cfg_simplify!(ir::IRCode)
                     if bb_rename_succ[catchbb] == 0
                         push!(worklist, catchbb)
                     end
+                elseif isa(terminator, GotoNode) || isa(terminator, ReturnNode)
+                    # No implicit fall through. Schedule from work list.
+                    break
+                else
+                    is_bottom = ir[SSAValue(bbs[curr].stmts[end])][:type] === Union{}
+                    if is_bottom && !isa(terminator, PhiNode) && terminator !== nothing
```

Comment:
I think this is missing an `Expr(:leave)` case above. Here's a case that seems to fail:
```julia
let code = Any[
               # block 1
               EnterNode(3, 1),
               # block 2
               GotoNode(4),
               # block 3
               ReturnNode(-1),
               # block 4
               Expr(:leave, SSAValue(1)),
               # block 5
               Expr(:call, :opaque),
               GotoNode(7),
               # block 6
               ReturnNode(nothing),
           ]
    ir = make_ircode(code; ssavaluetypes=Any[Union{} for i = 1:7])

    Core.Compiler.verify_ir(ir)
    ir = Core.Compiler.cfg_simplify!(ir)
    Core.Compiler.verify_ir(ir)
    @test length(ir.cfg.blocks) == 5 # should have merged block 2 into 4
end
```

Reply:
Expr(:leave) is not Bottom typed

Reply:
Can we add that to the verifier then?